samples: cpp: add a Hello, world! C++ app

cfriedt opened the pull request:
Create a separate hello world C++ app for Zephyr.

Commit:
samples: cpp: add a Hello, world! C++ app

Create a separate hello world C++ app for Zephyr.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

kartben commented:
/me likes! I'd say it probably belongs to the cpp/ samples folder though, as I'd prefer if each Hello World project (to your point, definitely to hope seeing other languages in the future) stays as naked as possible.

Unrelated: is the cpp synchronization sample still relevant in your opinion? Would be nice to refresh it with C11 Threads, no?

kartben commented on samples/hello_world/src/main.cpp:
```diff
@@ -0,0 +1,23 @@
+
+#include <stdio.h>
+
+class Greeter {
+public:
+	static void greet();
+};
+
+void Greeter::greet()
+{
+	printf("Hello C++ World! %s\n", CONFIG_BOARD);
+}
+
+int main(void)
+{
+	Greeter::greet();
+	return 0;
+}
```
I'd suggest sticking to the even more minimalistic "official" Hello World https://www.stroustrup.com/hello_world.c

cfriedt replied:
`<iostream>` is a beast - it's probably best to avoid for embedded / rtos purposes.

But maybe it works 🤷🏼‍♂️

cfriedt commented:
> /me likes! I'd say it probably belongs to the cpp/ samples folder though, as I'd prefer if each Hello World project (to your point, definitely to hope seeing other languages in the future) stays as naked as possible.

I can see the appeal of keeping languages fenced off, but that would be a lot of duplication for what is otherwise almost identical. I didn't think of updating the docs, but it would be nice to just cycle through a few different tabs on the same page to see how to accomplish the same trivial thing in a different language.

> Unrelated: is the cpp synchronization sample still relevant in your opinion? Would be nice to refresh it with C11 Threads, no?

Why settle for C11 threads and not C++11 threads? The SDK needs some patching, which is probably the biggest challenge (because it's very time intensive). But otherwise #43729 has been good to go for a long time.

gmarull reviewed (commented):
I'd personally create a separate sample, `hello_world_cpp` maybe

cfriedt commented:
> > I'd personally create a separate sample, `hello_world_cpp` maybe
> 
> Back to the the "tabs" idea though, having one actual sample per language certainly shouldn't prevent us from creating some kind of page cataloguing the various Zephyr Hello Worlds (eh, there might even be some out-of-tree ones worth referencing?)

That would work too. It's well beyond my level of .rst / sphinx, but maybe later on down the road. 

I'll change this so that it targets the cpp samples subdir

kartben commented on samples/cpp/hello_world/src/main.cpp:
```diff
@@ -0,0 +1,25 @@
+
+void Greeter::greet()
+{
+	/* note: iostream / std::cout are not used here due to size constraints */
```
👍🏼

cfriedt commented:
I have another rev to push. Would like to see if it's possible to use iostream with newlib

cfriedt commented on samples/hello_world/sample.yaml:
```diff
@@ -5,7 +5,7 @@ sample:
 common:
   tags: introduction
   integration_platforms:
-    - native_sim
```
because copy-paste - I'm using `qemu_riscv32` now.

cfriedt commented:
I haven't done a massive build for every platform yet. Hopefully nothing too serious.

cfriedt commented on samples/cpp/hello_world/sample.yaml:
```diff
@@ -0,0 +1,51 @@
+    type: one_line
+    regex:
+      - "Hello C\\+\\+, World! (.*)"
+tests:
```
Did you read the README.txt? It's explained there.

nashif replied:
Yes I did and I do not see how this is related. The sample demonstrates a c++ hello world with 2 confgiurations, so the most we need here are two tests that verify those are built correctly, there is no need for all the variations of c libraries. 
Just to give you an idea, for something as basic as this it triggers 100s of builds in CI and for verifying that a sample still builds and runs, it is way too much. If we are to test c++ support with different c libraries it needs to be done in a dedicated test, not a sample.

cfriedt replied:
> Just to give you an idea, for something as 

Why don't you suggest some specific changes?

stephanosio replied:
Given that this is a sample (i.e. not a test), I would say only demonstrating the default libcpp+libc should be sufficient.

After all, all the other libc+libcpp permutations should already be covered by other libcpp *tests*.

nashif replied:
> > Just to give you an idea, for something as
> 
> Why don't you suggest some specific changes?

I did already, read my comment:

The sample demonstrates a c++ hello world with 2 configurations, so the most we need here are two tests that verify those are built correctly, there is no need for all the variations of c libraries.

cfriedt commented:
It appears as though the iostream approach causes twister to fail on virtually all platforms even though the correct output appears on the terminal. Maybe it isn't flushed immediately or something along those lines. It would be strange if there were unusual characters being written to the output, but that doesn't appear to be the case.

yperess commented on samples/cpp/hello_world/src/main.cpp:
```diff
@@ -0,0 +1,13 @@
+
+int main(void)
+{
+	std::cout << "Hello C++, world! " << CONFIG_BOARD << std::endl;
```
Does this get routed to printk?

cfriedt replied:
Not 100% sure, but I feel like it should Just Work™. @stephanosio - any idea?

cfriedt replied:
> AFAIK std::endl includes a flush.
> That's why you should not use it to much 😅
> 
> Isn't iostream bloating? The last time I tried iostream added some 80k because of locale.

Yea, exactly. That's why I originally avoided it in the first rev of this PR. C++ hello world could just `#include` C hello world. I further expanded the original hello world so that you could add arbitrary languages with a couple lines of code, which was (IMHO) kind of short and sweet and to the point.

Then someone said it should be in a different project.

Then someone requested iostream, and so I did both iostream and printf versions.

Then someone said it introduced too much test complexity.

Then someone said it introduced too much build complexity.

In the end, I thought about just leaving a comment, and then realized that my editorial isn't even really necessary. Users can just figure out iostream is bloaty from existing sources and / or experimentation.

Then again, maybe someday someone will make a small-ish implementation of iostream (probably already exists - e.g. pigweed).

Future / not my problem.

nashif replied:
> > AFAIK std::endl includes a flush.
> > That's why you should not use it to much 😅
> > Isn't iostream bloating? The last time I tried iostream added some 80k because of locale.
> 
> Yea, exactly. That's why I originally avoided it in the first rev of this PR. C++ hello world could just `#include` C hello world. I further expanded the original hello world so that you could add arbitrary languages with a couple lines of code, which was (IMHO) kind of short and sweet and to the point.
> 
> Then someone said it should be in a different project.
> 
> Then someone requested iostream, and so I did both iostream and printf versions.
> 
> Then someone said it introduced too much test complexity.
> 
> Then someone said it introduced too much build complexity.
> 
> In the end, I thought about just leaving a comment, and then realized that my editorial isn't even really necessary. Users can just figure out iostream is bloaty from existing sources and / or experimentation.
> 
> Then again, maybe someday someone will make a small-ish implementation of iostream (probably already exists - e.g. pigweed).
> 
> Future / not my problem.

Hey, this just shows a working review process, even for something as simple as a hello world.

The iostream thing you could have dismissed as feature bloat and continued with a simple implementation, I did not see this being a firm request.

cfriedt replied:
> The iostream thing you could have dismissed as feature bloat and continued with a simple implementation, I did not see this being a firm request.

https://github.com/zephyrproject-rtos/zephyr/pull/65769#discussion_r1405249361

@nashif -

cfriedt replied:
A regex comma was misplaced 😅

cfriedt commented:
Works on my desk. For excluded platforms, patches welcome. 
<img width="867" alt="Screenshot 2023-11-30 at 7 38 31 AM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/afec8c32-5170-445a-adca-8e7856dcda20">

cfriedt commented:
@aescolar - hopefully this aligns with the goals of #65872 - I exclude `native_posix` and suggest `native_sim` in yaml comments.

cfriedt commented:
Still need to disable `native_posix` unfortunately, @aescolar, but I made your other requested changes.
```
ERROR   - native_posix              samples/cpp/hello_world/sample.cpp.helloworld       FAILED : Timeout
INFO    - /home/cfriedt/zephyrproject/zephyr/twister-out/native_posix/samples/cpp/hello_world/sample.cpp.helloworld/handler.log
ERROR   - *** Booting Zephyr OS build zephyr-v3.5.0-1864-gf13c2bd3ca3a ***

INFO    - /home/cfriedt/zephyrproject/zephyr/twister-out/native_posix/samples/cpp/hello_world/sample.cpp.helloworld/handler.log
INFO    - Total complete:   40/  40  100%  skipped:   19, failed:    1, error:    0
```

kartben commented on samples/cpp/hello_world/src/main.cpp:
```diff
@@ -0,0 +1,13 @@
+
+int main(void)
+{
+	std::cout << "Hello, C++ world! " << CONFIG_BOARD << std::endl;
```
not a request for change but a comment: I know this was discussed before so if you have evidence that people should avoid `std::cout` in real-life, might be worth mentioning? FWIW it does not seem to really bloat the ROM and RAM footprint at all compared to a simple printk, so there's that I guess.

cfriedt replied:
Mentioned before, but there is plenty of literature out there that says to avoid using `iostream` in an embedded environment. I don't need to repeat it here.

Also, that would preclude the use case where something like e.g. pigweed is used as a size-optimized replacement (assuming that can be done relatively easily by include path).

I think it's nice too, to simply support the de-facto C++ hello world (if we can) because it lowers the barrier to entry by using the standard approach, which is mostly the goal of hello world.

kartben reviewed (dismissed):
Very cool, thanks!

aescolar commented:
> Still need to disable `native_posix` unfortunately, @aescolar, but I made your other requested changes.

@cfriedt It does not sound right.
The PR as it is now works fine with native_posix* here :
``` 
zephyr$ twister -p native_sim -p native_sim_64 -p native_posix -p native_posix_64  -T samples/cpp/hello_world/ -v
Renaming output directory to /persistant/scratch/zephyrproject/zephyr/twister-out.3
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.5.0-1864-gf1f578704222
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /persistant/scratch/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - 1/4 native_sim                samples/cpp/hello_world/sample.cpp.helloworld      PASSED (native 2.612s)
INFO    - 2/4 native_sim_64             samples/cpp/hello_world/sample.cpp.helloworld      PASSED (native 2.611s)
INFO    - 3/4 native_posix_64           samples/cpp/hello_world/sample.cpp.helloworld      PASSED (native 2.611s)
INFO    - 4/4 native_posix              samples/cpp/hello_world/sample.cpp.helloworld      PASSED (native 2.612s)
``` 
Just enabling it
```diff
diff --git a/samples/cpp/hello_world/sample.yaml b/samples/cpp/hello_world/sample.yaml
index d86d20caa0..ce29b0df04 100644
--- a/samples/cpp/hello_world/sample.yaml
+++ b/samples/cpp/hello_world/sample.yaml
@@ -22,5 +22,3 @@ tests:
       - intel_ish_5_4_1
       - intel_ish_5_6_0
       - intel_ish_5_8_0
-      # please use native_sim instead
-      - native_posix
```

kartben commented on samples/cpp/hello_world/README.rst:
```diff
@@ -0,0 +1,33 @@
+Overview
+********
+
+A simple C++ sample that can be used with all :ref:`supported board <boards>` and prints
```
> Well, it does say "all supported boards", so a simple interpretation of that would be to say that unsupported boards are unsupported.

but it links to the full list of Zephyr supported boards, so I am not sure how one -- especially people who are the audience for such a Hello World sample, i.e more than likely Zephyr beginners -- would be supposed to know that there are, in fact, unsupported boards?

cfriedt replied:
I created #66027 to track the xtensa issue and zephyrproject-rtos/sdk-ng#593 actually existed from a long time ago when I tried to create a simple hello_world app too. I will make a note in those tickets to adjust the `sample.yaml` file when the fix is complete.

cfriedt replied:
Personally though, I would rather not revisit this application since it's so trivial, and I don't think we need to include bug reports (similarly to cmake internals and platform specific workarounds) in something as basic as hello world.

kartben replied:
To clarify, I was asking for improving the wording in the README, not the sample.yaml which I would assume a vast majority of the audience for "Hello World" kind of samples don't even know exist. If I am getting started with Zephyr and giving the C++ Hello World a try, there is a very high chance I will only look at the README initially, so if it says "this sample can be used on all supported boards", it is really confusing. And some people might not even read the README the at all, but that's where they will probably refer to in case it doesn't work, to check what they may have done wrong.

FWIW I agree that hopefully things will be fixed by the time we cut a release, but in the meantime (or in the event that it does *not* get fixed before the release), I really think the documentation should point out any potential catch or limitation.

cfriedt commented:
The PR is here for reference. Please do whatever you like with it @kartben.

cfriedt commented:
> Thanks. I will have a look and try to push it over the finish line. To clarify I was mainly/only asking that the wording "all supported boards" is not used in combination with an hyperlink taking one to the list of all Zephyr's supported boards, as this can be quite error-prone to newcomers who wouldn't understand why their board is in fact not (currently) working.

Done! I reworded to say "many supported boards", removed the hyperlink, and rebased.

Does that work @kartben?

kartben commented on samples/cpp/hello_world/README.rst:
```diff
@@ -0,0 +1,33 @@
+Overview
+********
+
+A simple :ref:`C++ <language_cpp>` sample that can be used with many supported board and prints
```
```suggestion
A simple :ref:`C++ <language_cpp>` sample that can be used with many supported boards and prints
```

cfriedt replied:
Fixed!
